### PR TITLE
Updated build.fsx to update AssemblyInfo.cs

### PR DESCRIPTION
**Pull request by NogginBops**
### Purpose of this PR

Changes the build system to generate new `AssemblyInfo.cs` with updated version info.

I wasn't able to find a way to keep the `InternalsVisibleToAttributes` specified in `OpenTK.Core.csproj` without getting attribute conflicts. So ideally we would still retain the possibility of using this attribute.

I decided that it was time to delete `MarshalUtility.cs` as it has played its role and is no longer needed.

What we want to do with `MouseCursor.StandardShape` is not obvious, but for now I've decided it's fine to expose this enum. But maybe the solution is to merge `OpenTK.Windowing.Common` into the other projects as it doesn't really have a place any more.

### Testing status

This builds and produces nuget packages which have updated assembly versions.
I've not tested these nuget packages to see if they work as expected.

Changing build system stuff is scary, and I don't know if I've done things right, so any reviews or testing is highly appreciated.

### CODE REVIEW

File: build.fsx [hunk 1]
```diff
@@ -190,6 +190,8 @@ let rewriteBindsForTfm tfm =
     let args = [ "-a"; targetPath ] |> asArgs
     DotNet.runWithDefaultOptions framework projFile args |> ignore
 
+// FIXME: We want to propegate errors from rewriteBindsForTfm
```
Comment (varon) [inline build.fsx:193]: propegate -> propagate

File: build.fsx [hunk 2]
```diff
@@ -219,7 +221,43 @@ Target.create "Restore" (fun _ -> DotNet.restore dotnetSimple "OpenTK.sln" |> ig
...
     // see https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019#directorybuildprops-and-directorybuildtargets
-    Trace.traceError "Unimplemented.")
+
+    let getAssemblyInfoAttributes (projectName:string) =
+        [ AssemblyInfo.Title (projectName)
+          AssemblyInfo.Product project
+          AssemblyInfo.Description summary
+          AssemblyInfo.Version release.AssemblyVersion
+          AssemblyInfo.FileVersion release.AssemblyVersion
+          AssemblyInfo.CLSCompliant true
```
Comment (varon) [inline build.fsx:231]: I don't think we're technically CLS compliant anymore. This isn't really an attribute we (want) to provide, so let's set this to false. It's only really for VB.net users who aren't likely to be our target going forward.

File: build.fsx [hunk 3]
```diff
@@ -219,7 +221,43 @@ Target.create "Restore" (fun _ -> DotNet.restore dotnetSimple "OpenTK.sln" |> ig
...
+          AssemblyInfo.Copyright copyright
+        ]
+
+    let getProjectDetails (projectPath:string) =
+        let projectName = System.IO.Path.GetFileNameWithoutExtension(projectPath)
+        ( projectPath,
+          projectName,
+          System.IO.Path.GetDirectoryName(projectPath),
+          (getAssemblyInfoAttributes projectName)
+        )
```
Comment (varon) [inline build.fsx:241]: This is a good candidate for an anonymous record instead:
Try replace this with:
```fsharp
{|
    ProjectPath = projectPath
    ProjectName = projectName
    DirectoryName = System.IO.Path.GetDirectoryName(projectPath)
    AssemblyInfoAttributes = getAssemblyInfoAttributes projectName
|}
```

Then you can access the record fields by name later on for a bit more simplicity/clarity.

File: build.fsx [hunk 4]
```diff
@@ -219,7 +221,43 @@ Target.create "Restore" (fun _ -> DotNet.restore dotnetSimple "OpenTK.sln" |> ig
...
+        | _                           -> failwith (sprintf "Project file %s not supported. Unknown project type." projFileName)
+
+    releaseProjects
+    |> Seq.map getProjectDetails
+    |> Seq.iter (fun (projFileName, projectName, folderName, attributes) ->
+        match projFileName with
+        | Fsproj -> AssemblyInfoFile.createFSharp (folderName @@ "AssemblyInfo.fs") attributes
+        | Csproj -> AssemblyInfoFile.createCSharp ((folderName @@ "Properties") @@ "AssemblyInfo.cs") attributes
+        | Vbproj -> AssemblyInfoFile.createVisualBasic ((folderName @@ "My Project") @@ "AssemblyInfo.vb") attributes
+        )
```
Comment (varon) [inline build.fsx:258]: This we can replace with the anonymous record instead.

File: src/OpenTK.Core/Utility/MarshalUtility.cs [hunk 5]
```diff
@@ -1,23 +0,0 @@
-using System;
-using System.Runtime.InteropServices;
-
-namespace OpenTK.Core
-{
-    internal static class MarshalUtility
```
Comment (varon) [inline src/OpenTK.Core/Utility/MarshalUtility.cs:6]: Removal of this is a good change.

**Review by varon — COMMENTED** [review]
Looks good - just some minor conceptual stuff we want to improve on for the records/tuple usage.

For the assembly info, in general that's largely supplanted by the Nuget package version instead. A lot of assemblies are just sitting on 1.0.0 for all eternity and letting the versioning/metadata be handled by the Nuget package instead.

If we're sure we want to do that, then these changes definitely accomplish that.

**Comment by utkumaden** [comment]
I tried verifying the build on my own Linux machine by running `build.sh` and I seem to get a build error, however I am unsure if it is relevant to the changes in the PR. It seems to not break the CI build so it could just be an "on my machine" thing.
```
Exec: /tmp/opentk/src/Generator/Parsing.fs(89,10): error FS0039: The type 'Registry' does not define the field, constructor or member 'Groups'. [/tmp/opentk/src/Generator/Generator.fsproj]
Exec: /tmp/opentk/src/Generator/Parsing.fs(92,25): error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved. [/tmp/opentk/src/Generator/Generator.fsproj]
Exec: /tmp/opentk/src/Generator/Parsing.fs(95,17): error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved. [/tmp/opentk/src/Generator/Generator.fsproj]
Exec: /tmp/opentk/src/Generator/Parsing.fs(98,36): error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved. [/tmp/opentk/src/Generator/Generator.fsproj]
Exec: /tmp/opentk/src/Generator/Parsing.fs(100,34): error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to constrain the type of the object. This may allow the lookup to be resolved. [/tmp/opentk/src/Generator/Generator.fsproj]
Exec: /tmp/opentk/src/OpenTK.Graphics/OpenTK.Graphics.csproj(22,5): error MSB3073: The command "bash build.sh -t UpdateBindings" exited with code -1.
```

**Comment by NogginBops** [comment]
It's weird because it says that it's running `bash build.sh -t UpdateBindings` which is the one target that should not be run in the build.fsx. Maybe if you remove those lines in the `build.fsx`. @utkumaden 
https://github.com/opentk/opentk/blob/master/build.fsx#L163-L172

**Comment by NogginBops** [comment]
I will not merge this PR. It's not worth having this change to the build system if no one is going to benefit from the changes.

**Pull request closed without merging**